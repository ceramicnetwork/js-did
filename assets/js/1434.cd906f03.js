"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[1434],{1434:(e,s,b)=>{b.r(s)}}]);