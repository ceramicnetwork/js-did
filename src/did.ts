import { ResolverOptions, DIDResolutionResult, ResolverRegistry, Resolver } from 'did-resolver'
import { createJWE, JWE, verifyJWS, resolveX25519Encrypters } from 'did-jwt'
import { encodePayload, prepareCleartext, decodeCleartext } from 'dag-jose-utils'
import { RPCClient } from 'rpc-utils'
import { CID } from 'multiformats/cid'
import { CacaoBlock, Cacao } from 'ceramic-cacao'
import type { DagJWS, DIDProvider, DIDProviderClient } from './types.js'
import {
  fromDagJWS,
  encodeBase64,
  base64urlToJSON,
  decodeBase64,
  encodeBase64Url,
  randomString,
  didWithTime,
  extractControllers,
} from './utils.js'

export interface AuthenticateOptions {
  provider?: DIDProvider
  aud?: string
  paths?: Array<string>
}

export interface AuthenticateParams {
  nonce: string
  aud?: string
  paths?: Array<string>
}

export interface AuthenticateResponse extends AuthenticateParams {
  did: string
  exp: number
}

export interface CreateJWSOptions {
  did?: string
  protected?: Record<string, any>
  linkedBlock?: string
}

export interface VerifyJWSOptions {
  /**
   * JS timestamp when the signature was allegedly made. `undefined` means _now_.
   */
  atTime?: Date

  /**
   * If true, timestamp checking is disabled.
   */
  disableTimecheck?: boolean

  /**
   * DID that issued the signature.
   */
  issuer?: string

  /**
   * Cacao OCAP to verify the JWS with.
   */
  capability?: Cacao

  /**
   * Number of seconds that a revoked key stays valid for after it was revoked
   */
  revocationPhaseOutSecs?: number
}

export interface VerifyJWSResult {
  kid: string
  payload?: Record<string, any>
  didResolutionResult: DIDResolutionResult
}

export interface CreateJWEOptions {
  protectedHeader?: Record<string, any>
  aad?: Uint8Array
}

export interface DecryptJWEOptions {
  did?: string
}

export interface DecryptJWEResult {
  cleartext: string // base64-encoded
}

export interface DagJWSResult {
  jws: DagJWS
  linkedBlock: Uint8Array
  cacaoBlock?: Uint8Array
}

export interface DIDOptions {
  provider?: DIDProvider
  resolver?: Resolver | ResolverRegistry
  resolverOptions?: ResolverOptions
  capability?: Cacao
}

function isResolver(resolver: Resolver | ResolverRegistry): resolver is Resolver {
  return 'registry' in resolver && 'cache' in resolver
}

/**
 * Interact with DIDs.
 */
export class DID {
  private _client?: DIDProviderClient
  private _id?: string
  private _resolver!: Resolver
  private _capability?: Cacao

  constructor({ provider, resolver = {}, resolverOptions, capability }: DIDOptions = {}) {
    if (provider != null) {
      this._client = new RPCClient(provider)
    }
    if (capability) {
      this._capability = capability
    }
    this.setResolver(resolver, resolverOptions)
  }

  /**
   * Check if the DID has a capability attached
   */
  get capability(): Cacao {
    if (!this._capability) {
      throw new Error('DID has no capability attached')
    }
    return this._capability
  }

  /**
   * Check if user is authenticated.
   */
  get authenticated(): boolean {
    return this._id != null
  }

  /**
   * Get the DID identifier of the user.
   */
  get id(): string {
    if (this._id == null) {
      throw new Error('DID is not authenticated')
    }
    return this._id
  }

  /**
   * Attach a capability to the DID instance
   * @param cap The capability to attach
   * @returns A new DID instance with the capability attached
   */
  withCapability(cap: Cacao): DID {
    return new DID({
      provider: this._client?.connection,
      resolver: this._resolver,
      capability: cap,
    })
  }

  /**
   * Set the DID provider of this instance.
   * Only callable if provider not already set.
   *
   * @param provider    The DIDProvider to use
   */
  setProvider(provider: DIDProvider): void {
    if (this._client == null) {
      this._client = new RPCClient(provider)
    } else if (this._client.connection !== provider) {
      throw new Error(
        'A different provider is already set, create a new DID instance to use another provider'
      )
    }
  }

  /**
   * Set the DID-resolver user by this instance
   *
   * @param resolver            Either a Resolver instance or an object with specific resolvers
   * @param resolverOptions     Options to use for the created resolver. Will be ignored if a Resolver instance is passed
   */
  setResolver(resolver: Resolver | ResolverRegistry, resolverOptions?: ResolverOptions): void {
    this._resolver = isResolver(resolver) ? resolver : new Resolver(resolver, resolverOptions)
  }

  /**
   * Authenticate the user.
   */
  async authenticate({ provider, paths = [], aud }: AuthenticateOptions = {}): Promise<string> {
    if (provider != null) {
      this.setProvider(provider)
    }
    if (this._client == null) {
      throw new Error('No provider available')
    }
    const nonce = randomString()
    const jws = await this._client.request('did_authenticate', {
      nonce,
      aud,
      paths,
    })
    const { kid } = await this.verifyJWS(jws)
    const payload = base64urlToJSON(jws.payload) as AuthenticateResponse
    if (!kid.includes(payload.did)) throw new Error('Invalid authencation response, kid mismatch')
    if (payload.nonce !== nonce) throw new Error('Invalid authencation response, wrong nonce')
    if (payload.aud !== aud) throw new Error('Invalid authencation response, wrong aud')
    if (payload.exp < Date.now() / 1000) throw new Error('Invalid authencation response, expired')
    this._id = payload.did
    return this._id
  }

  /**
   * Create a JWS encoded signature over the given payload.
   * Will be signed by the currently authenticated DID.
   *
   * @param payload             The payload to sign
   * @param options             Optional parameters
   */
  async createJWS<T = any>(payload: T, options: CreateJWSOptions = {}): Promise<DagJWS> {
    if (this._client == null) throw new Error('No provider available')
    if (this._id == null) throw new Error('DID is not authenticated')
    if (this._capability) {
      const cacaoBlock = await CacaoBlock.fromCacao(this._capability)
      const capCID = CID.asCID(cacaoBlock.cid)
      if (!capCID) {
        throw new Error(
          `Capability CID of the JWS cannot be set to the capability payload cid as they are incompatible`
        )
      }
      options.protected = options.protected || {}
      options.protected.cap = `ipfs://${capCID?.toString()}`
    }
    const { jws } = await this._client.request('did_createJWS', {
      did: this._id,
      ...options,
      payload,
    })
    return jws
  }

  /**
   * Create an IPFS compatibe DagJWS encoded signature over the given payload.
   * Will be signed by the currently authenticated DID.
   *
   * @param payload             The payload to sign, may include ipld links
   * @param options             Optional parameters
   */
  async createDagJWS(
    payload: Record<string, any>,
    options: CreateJWSOptions = {}
  ): Promise<DagJWSResult> {
    const { cid, linkedBlock } = await encodePayload(payload)
    const payloadCid = encodeBase64Url(cid.bytes)
    Object.assign(options, { linkedBlock: encodeBase64(linkedBlock) })
    const jws = await this.createJWS(payloadCid, options)

    const compatibleCID = CID.asCID(cid)
    if (!compatibleCID) {
      throw new Error(
        'CID of the JWS cannot be set to the encoded payload cid as they are incompatible'
      )
    }
    jws.link = compatibleCID

    if (this._capability) {
      const cacaoBlock = await CacaoBlock.fromCacao(this._capability)
      return { jws, linkedBlock, cacaoBlock: cacaoBlock.bytes }
    }
    return { jws, linkedBlock }
  }

  /**
   * Verify a JWS. Uses the 'kid' in the header as the way to resolve
   * the author public key.
   *
   * @param jws                 The JWS to verify
   * @param options             Optional arguments for verification
   * @returns                   Information about the signed JWS
   */
  async verifyJWS(jws: string | DagJWS, options: VerifyJWSOptions = {}): Promise<VerifyJWSResult> {
    if (typeof jws !== 'string') jws = fromDagJWS(jws)
    const kid = base64urlToJSON(jws.split('.')[0]).kid as string
    if (!kid) throw new Error('No "kid" found in jws')
    const didResolutionResult = await this.resolve(kid)
    const timecheckEnabled = !options.disableTimecheck
    if (timecheckEnabled) {
      const nextUpdate = didResolutionResult.didDocumentMetadata?.nextUpdate
      if (nextUpdate) {
        // This version of the DID document has been revoked. Check if the JWS
        // was signed before the revocation happened.
        const phaseOutMS = options.revocationPhaseOutSecs
          ? options.revocationPhaseOutSecs * 1000
          : 0
        const revocationTime = new Date(nextUpdate).valueOf() + phaseOutMS
        const isEarlier = options.atTime && options.atTime.getTime() < revocationTime
        const isLater = !isEarlier
        if (isLater) {
          // Do not allow using a key _after_ it is being revoked
          throw new Error(`invalid_jws: signature authored with a revoked DID version: ${kid}`)
        }
      }
      // Key used before `updated` date
      const updated = didResolutionResult.didDocumentMetadata?.updated
      if (updated && options.atTime && options.atTime.getTime() < new Date(updated).valueOf()) {
        throw new Error(`invalid_jws: signature authored before creation of DID version: ${kid}`)
      }
    }

    const signerDid = didResolutionResult.didDocument?.id
    if (
      options.issuer &&
      options.issuer === options.capability?.p.iss &&
      signerDid === options.capability.p.aud
    ) {
      Cacao.verify(options.capability, {
        atTime: options.atTime ? options.atTime : undefined,
      })
    } else if (options.issuer && options.issuer !== signerDid) {
      const issuerUrl = didWithTime(options.issuer, options.atTime)
      const issuerResolution = await this.resolve(issuerUrl)
      const controllerProperty = issuerResolution.didDocument?.controller
      const controllers = extractControllers(controllerProperty)

      if (
        options.capability?.s &&
        options.capability.p.aud === signerDid &&
        controllers.includes(options.capability.p.iss)
      ) {
        Cacao.verify(options.capability, {
          atTime: options.atTime ? options.atTime : undefined,
        })
      } else {
        const signerIsController = signerDid ? controllers.includes(signerDid) : false
        if (!signerIsController) {
          throw new Error(`invalid_jws: not a valid verificationMethod for issuer: ${kid}`)
        }
      }
    }

    const publicKeys = didResolutionResult.didDocument?.verificationMethod || []
    // verifyJWS will throw an error if the signature is invalid
    verifyJWS(jws, publicKeys)
    let payload
    try {
      payload = base64urlToJSON(jws.split('.')[1])
    } catch (e) {
      // If an error is thrown it means that the payload is a CID.
    }
    return { kid, payload, didResolutionResult }
  }

  /**
   * Create a JWE encrypted to the given recipients.
   *
   * @param cleartext           The cleartext to be encrypted
   * @param recipients          An array of DIDs
   * @param options             Optional parameters
   */
  async createJWE(
    cleartext: Uint8Array,
    recipients: Array<string>,
    options: CreateJWEOptions = {}
  ): Promise<JWE> {
    const encrypters = await resolveX25519Encrypters(recipients, this._resolver)
    return createJWE(cleartext, encrypters, options.protectedHeader, options.aad)
  }

  /**
   * Create an IPFS compatibe DagJWE encrypted to the given recipients.
   *
   * @param cleartext           The cleartext to be encrypted, may include ipld links
   * @param recipients          An array of DIDs
   * @param options             Optional parameters
   */
  async createDagJWE(
    cleartext: Record<string, any>,
    recipients: Array<string>,
    options: CreateJWEOptions = {}
  ): Promise<JWE> {
    const preparedCleartext = await prepareCleartext(cleartext)
    return this.createJWE(preparedCleartext, recipients, options)
  }

  /**
   * Try to decrypt the given JWE with the currently authenticated user.
   *
   * @param jwe                 The JWE to decrypt
   * @param options             Optional parameters
   */
  async decryptJWE(jwe: JWE, options: DecryptJWEOptions = {}): Promise<Uint8Array> {
    if (this._client == null) throw new Error('No provider available')
    if (this._id == null) throw new Error('DID is not authenticated')
    const { cleartext } = await this._client.request('did_decryptJWE', {
      did: this._id,
      ...options,
      jwe,
    })
    return decodeBase64(cleartext)
  }

  /**
   * Try to decrypt the given DagJWE with the currently authenticated user.
   *
   * @param jwe                 The JWE to decrypt
   * @returns                   An ipld object
   */
  async decryptDagJWE(jwe: JWE): Promise<Record<string, any>> {
    const bytes = await this.decryptJWE(jwe)
    return decodeCleartext(bytes)
  }

  /**
   * Resolve the DID Document of the given DID.
   *
   * @param didUrl              The DID to resolve
   */
  async resolve(didUrl: string): Promise<DIDResolutionResult> {
    const result = await this._resolver.resolve(didUrl)
    if (result.didResolutionMetadata.error) {
      const { error, message } = result.didResolutionMetadata
      const maybeMessage = message ? `, ${message as string}` : ''
      throw new Error(`Failed to resolve ${didUrl}: ${error}${maybeMessage}`)
    }
    return result
  }
}
