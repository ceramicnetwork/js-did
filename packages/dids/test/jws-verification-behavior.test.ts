import { jest } from '@jest/globals'
import { DID } from '../src/did.js'

const VERSION_0_VANILLA = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
    verificationMethod: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#7wYNHm3nGoNA3Kv',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: '2BS946XE3ZGDEsFSq5G3ndodLnAz2TZu1SN1FUw8EwZUC',
      },
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#ATBW8KGCmgnXaKR',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'HZCZPM7PzCGow5WxopEKeQaKNGQgsZH2d9baiK3FpCYf',
      },
    ],
    authentication: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#7wYNHm3nGoNA3Kv',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: '2BS946XE3ZGDEsFSq5G3ndodLnAz2TZu1SN1FUw8EwZUC',
      },
    ],
    keyAgreement: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#ATBW8KGCmgnXaKR',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'HZCZPM7PzCGow5WxopEKeQaKNGQgsZH2d9baiK3FpCYf',
      },
    ],
  },
  didDocumentMetadata: {
    versionId: '0',
  },
}

const VERSION_0_ROTATED = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
    verificationMethod: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#7wYNHm3nGoNA3Kv',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: '2BS946XE3ZGDEsFSq5G3ndodLnAz2TZu1SN1FUw8EwZUC',
      },
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#ATBW8KGCmgnXaKR',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'HZCZPM7PzCGow5WxopEKeQaKNGQgsZH2d9baiK3FpCYf',
      },
    ],
    authentication: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#7wYNHm3nGoNA3Kv',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: '2BS946XE3ZGDEsFSq5G3ndodLnAz2TZu1SN1FUw8EwZUC',
      },
    ],
    keyAgreement: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#ATBW8KGCmgnXaKR',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'HZCZPM7PzCGow5WxopEKeQaKNGQgsZH2d9baiK3FpCYf',
      },
    ],
  },
  didDocumentMetadata: {
    nextUpdate: '2021-07-07T08:12:19Z',
    versionId: '0',
    nextVersionId: 'bafyreiebh2rwia4qq5v43dqlnofkczsvoghdxymahv4p5lkp5ivr5kgatm',
  },
}

const VERSION_NEXT = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
    verificationMethod: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#C26y2DZeJ9EMYsF',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'vTcTTxzsSsAiqTb9EH7tC13gG1Dfkdya36TmLxU7951X',
      },
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#sfecbasdBzBfhcb',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'H1s1bgxnxZPdhF2ZXqACwi3QxPCsZGVVjcsC8jLexKqq',
      },
    ],
    authentication: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#C26y2DZeJ9EMYsF',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'vTcTTxzsSsAiqTb9EH7tC13gG1Dfkdya36TmLxU7951X',
      },
    ],
    keyAgreement: [
      {
        id: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k#sfecbasdBzBfhcb',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k',
        publicKeyBase58: 'H1s1bgxnxZPdhF2ZXqACwi3QxPCsZGVVjcsC8jLexKqq',
      },
    ],
  },
  didDocumentMetadata: {
    updated: '2021-07-07T08:12:19Z',
    versionId: 'bafyreiebh2rwia4qq5v43dqlnofkczsvoghdxymahv4p5lkp5ivr5kgatm',
  },
}

const COMPOSITE_ISSUER_EMPTY = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:composite:foo',
    controller: [],
  },
  didDocumentMetadata: {},
}

// v0: did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=0#7wYNHm3nGoNA3Kv
const jwsV0 =
  'eyJraWQiOiJkaWQ6MzpranpsNmN3ZTFqdzE0YWg4d2p5OGdyZ3psNTJzbDE4c2J5aXJnaTlicXk5eXp1MjhrYnh2am1oaXA5OXIxNGs_dmVyc2lvbi1pZD0wIzd3WU5IbTNuR29OQTNLdiIsImFsZyI6IkVTMjU2SyJ9.eyJoZWxsbyI6IndvcmxkIn0.cHVXU7QW2pvBJpVIBCLhnkCQ0k4Up3cNRqiyeryRbPOSrZdAoQmWy1OfzNFzgY90nol26KJxHWnknSyu5sY__Q'
// vNext: did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=bafyreiebh2rwia4qq5v43dqlnofkczsvoghdxymahv4p5lkp5ivr5kgatm#C26y2DZeJ9EMYsF
const jwsVNext =
  'eyJraWQiOiJkaWQ6MzpranpsNmN3ZTFqdzE0YWg4d2p5OGdyZ3psNTJzbDE4c2J5aXJnaTlicXk5eXp1MjhrYnh2am1oaXA5OXIxNGs_dmVyc2lvbi1pZD1iYWZ5cmVpZWJoMnJ3aWE0cXE1djQzZHFsbm9ma2N6c3ZvZ2hkeHltYWh2NHA1bGtwNWl2cjVrZ2F0bSNDMjZ5MkRaZUo5RU1Zc0YiLCJhbGciOiJFUzI1NksifQ.eyJoZWxsbyI6IndvcmxkIn0.GRKdQMwbGw5t8Q_8asCV88jofZhJaPZlLs82-E_DvNLwfRBR8_Sbt_GwHPqdghA9oSN3XrTsI9KlKjwf8G6sew'

test('vanilla version 0', async () => {
  const did = new DID()
  did.resolve = () => Promise.resolve(VERSION_0_VANILLA)
  const { kid } = await did.verifyJWS(jwsV0)
  expect(kid).toEqual(
    'did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=0#7wYNHm3nGoNA3Kv'
  )
})

describe('rotatedKey', () => {
  const did = new DID()
  did.resolve = (didUrl) => {
    const isVersion0 = /version-id=0/.exec(didUrl)
    if (isVersion0) {
      return Promise.resolve(VERSION_0_ROTATED)
    } else {
      return Promise.resolve(VERSION_NEXT)
    }
  }

  test('throw', async () => {
    await expect(did.verifyJWS(jwsV0)).rejects.toThrow(
      /invalid_jws: signature authored with a revoked DID version/
    )
  })

  test('pass if timecheck disabled', async () => {
    await expect(did.verifyJWS(jwsV0, { disableTimecheck: true })).resolves.toBeTruthy()
  })
})

describe('atTime', () => {
  const did = new DID()
  const fauxResolve = (didUrl: string) => {
    const isVersion0 = /version-id=0/.exec(didUrl)
    if (isVersion0) {
      return Promise.resolve(VERSION_0_ROTATED)
    } else {
      return Promise.resolve(VERSION_NEXT)
    }
  }
  beforeEach(() => {
    did.resolve = fauxResolve
  })

  const beforeRotation = new Date('2021-07-07T08:00:19Z')
  const afterRotation = new Date('2021-07-07T08:40:19Z')
  const timeKeysRotatedInSec = Math.floor(
    new Date(VERSION_0_ROTATED.didDocumentMetadata.nextUpdate).valueOf() / 1000
  )

  test('ok before rotation', async () => {
    const { kid } = await did.verifyJWS(jwsV0, { atTime: beforeRotation })
    expect(kid).toMatchSnapshot()
  })
  test('fail after rotation', async () => {
    await expect(did.verifyJWS(jwsV0, { atTime: afterRotation })).rejects.toThrow(
      /invalid_jws: signature authored with a revoked DID version/
    )
  })
  test('ok after rotation if timecheck disabled', async () => {
    await expect(
      did.verifyJWS(jwsV0, { atTime: afterRotation, disableTimecheck: true })
    ).resolves.toBeTruthy()
  })
  test('ok after rotation if within revocation phase out period', async () => {
    // values in s
    const revocationPhaseOutSecs = 10
    const afterRotationBeforePhaseOut = timeKeysRotatedInSec + revocationPhaseOutSecs - 1

    await expect(
      did.verifyJWS(jwsV0, {
        atTime: new Date(afterRotationBeforePhaseOut * 1000),
        revocationPhaseOutSecs,
      })
    ).resolves.toBeTruthy()
  })
  test('fail after rotation if not within revocation phase out period', async () => {
    // values in s
    const revocationPhaseOutSecs = 10
    const afterRotationAfterPhaseOut = timeKeysRotatedInSec + revocationPhaseOutSecs

    await expect(
      did.verifyJWS(jwsV0, {
        atTime: new Date(afterRotationAfterPhaseOut * 1000),
        revocationPhaseOutSecs,
      })
    ).rejects.toThrow(/invalid_jws: signature authored with a revoked DID version/)
  })

  test('before DID version available', async () => {
    did.resolve = () => Promise.resolve(VERSION_NEXT)
    await expect(did.verifyJWS(jwsVNext, { atTime: beforeRotation })).rejects.toThrow(
      /invalid_jws: signature authored before creation/
    )
  })
  test('before DID version available if timecheck disabled', async () => {
    did.resolve = () => Promise.resolve(VERSION_NEXT)
    // Have lost keys for the old DID, so here is a new one.
    await expect(
      did.verifyJWS(jwsVNext, { atTime: beforeRotation, disableTimecheck: true })
    ).resolves.toBeTruthy()
  })
  test('before DID available for v0', async () => {
    const { kid } = await did.verifyJWS(jwsV0, { atTime: beforeRotation })
    expect(kid).toMatchSnapshot()
  })
})

describe('issuer', () => {
  const SIGNER_DID = VERSION_0_VANILLA.didDocument.id
  const did = new DID()

  beforeEach(() => {
    did.resolve = jest.fn((didUrl: string) => {
      if (didUrl === COMPOSITE_ISSUER_EMPTY.didDocument.id) {
        return Promise.resolve(COMPOSITE_ISSUER_EMPTY)
      }
      return Promise.resolve(VERSION_0_VANILLA)
    })
  })

  test('same as signer', async () => {
    const { kid } = await did.verifyJWS(jwsV0, { issuer: SIGNER_DID })
    expect(kid).toMatchSnapshot()
  })
  test('includes signer as controller', async () => {
    const fauxResolve = jest.fn((didUrl: string) => {
      if (didUrl === COMPOSITE_ISSUER_EMPTY.didDocument.id) {
        const included = {
          ...COMPOSITE_ISSUER_EMPTY,
          didDocument: {
            ...COMPOSITE_ISSUER_EMPTY.didDocument,
            controller: [SIGNER_DID],
          },
        }
        return Promise.resolve(included)
      }
      return Promise.resolve(VERSION_0_VANILLA)
    })
    did.resolve = fauxResolve
    const issuer = COMPOSITE_ISSUER_EMPTY.didDocument.id
    const { kid } = await did.verifyJWS(jwsV0, { issuer: issuer })
    expect(kid).toMatchSnapshot()
    expect(fauxResolve).toBeCalledWith(issuer)
  })
  test('does not include signer as controller', async () => {
    const issuer = COMPOSITE_ISSUER_EMPTY.didDocument.id
    await expect(did.verifyJWS(jwsV0, { issuer: issuer })).rejects.toThrow(
      /invalid_jws: not a valid verificationMethod for issuer/
    )
  })
})
