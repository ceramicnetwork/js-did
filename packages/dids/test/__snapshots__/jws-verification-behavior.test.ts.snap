// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atTime before DID available for v0 1`] = `"did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=0#7wYNHm3nGoNA3Kv"`;

exports[`atTime ok before rotation 1`] = `"did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=0#7wYNHm3nGoNA3Kv"`;

exports[`issuer includes signer as controller 1`] = `"did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=0#7wYNHm3nGoNA3Kv"`;

exports[`issuer same as signer 1`] = `"did:3:kjzl6cwe1jw14ah8wjy8grgzl52sl18sbyirgi9bqy9yzu28kbxvjmhip99r14k?version-id=0#7wYNHm3nGoNA3Kv"`;
